File: web/src/kernel/validation.ts
/**
 * Shared receipt validation levels for Langarian Math Workbench v0.3
 * (TS mirror of src/langarian/validation.py).
 *
 * Four distinct, separately reported levels:
 * - schema: required fields, types, known enum values, >= 1 invariant
 *   result. Passing this level alone is a shape-only check and must never
 *   be labeled "verified".
 * - hash: content_hash and receipt_id recomputed from the body must match
 *   the recorded values (tamper detection).
 * - status: recorded status must equal the status recomputed by collapsing
 *   invariant_results (empty invariants => FAIL; FAILED tag override).
 * - version: kernel/metric/receipt-schema versions must be in the current
 *   allowlists; older or unknown versions are rejected.
 *
 * Even full success is local consistency verification, not recomputation of
 * the underlying mathematical operation.
 */

import { canonicalJson, type CanonicalValue } from './canonical.js'
import { combineStatuses, type ResultStatus } from './contracts.js'
import { sha256Prefixed } from './sha256.js'
import {
  ALLOWED_KERNEL_VERSIONS,
  ALLOWED_METRIC_VERSIONS,
  ALLOWED_RECEIPT_SCHEMA_VERSIONS,
} from './version.js'

export const REQUIRED_RECEIPT_FIELDS: ReadonlySet<string> = new Set([
  'receipt_id',
  'content_hash',
  'kernel_version',
  'metric_version',
  'receipt_schema_version',
  'operator',
  'input_hashes',
  'output_hash',
  'invariant_results',
  'status',
  'epistemic_tag',
])
export const VALID_STATUSES: ReadonlySet<string> = new Set(['PASS', 'WARN', 'FAIL'])
export const VALID_TAGS: ReadonlySet<string> = new Set([
  'FORMAL',
  'COMPUTED',
  'MODEL',
  'INTERPRETIVE',
  'METAPHOR',
  'OBSERVED',
  'FAILED',
])

export const LEVEL_SCHEMA = 'schema'
export const LEVEL_HASH = 'hash'
export const LEVEL_STATUS = 'status'
export const LEVEL_VERSION = 'version'
export const LEVEL_ORDER = [LEVEL_SCHEMA, LEVEL_HASH, LEVEL_STATUS, LEVEL_VERSION] as const

const IDENTITY_FIELDS = new Set(['receipt_id', 'content_hash'])

export interface ValidationLevel {
  name: string
  ok: boolean
  errors: readonly string[]
}

export interface ValidationLevelSummary {
  schema_valid: boolean
  hash_valid: boolean
  status_consistent: boolean
  version_allowed: boolean
}

export class ReceiptValidation {
  readonly levels: readonly ValidationLevel[]

  constructor(levels: readonly ValidationLevel[]) {
    this.levels = levels
  }

  get ok(): boolean {
    return this.levels.every((level) => level.ok)
  }

  /** True when only the schema level passed — NOT verification. */
  get schemaOnlyOk(): boolean {
    return this.level(LEVEL_SCHEMA).ok && !this.ok
  }

  /** Named four-level summary used by the UI and conformance fixtures. */
  get summary(): ValidationLevelSummary {
    return {
      schema_valid: this.level(LEVEL_SCHEMA).ok,
      hash_valid: this.level(LEVEL_HASH).ok,
      status_consistent: this.level(LEVEL_STATUS).ok,
      version_allowed: this.level(LEVEL_VERSION).ok,
    }
  }

  level(name: string): ValidationLevel {
    const found = this.levels.find((level) => level.name === name)
    if (!found) throw new Error(`no such validation level: ${name}`)
    return found
  }

  errorsFlat(): string[] {
    return this.levels.flatMap((level) => level.errors.map((error) => `${level.name}: ${error}`))
  }

  toDict(): Record<string, CanonicalValue> {
    return {
      ok: this.ok,
      levels: this.levels.map((level) => ({
        name: level.name,
        ok: level.ok,
        errors: Array.from(level.errors),
      })),
    }
  }
}

type ReceiptData = Record<string, CanonicalValue>

function hashJson(data: ReceiptData): string {
  return sha256Prefixed(canonicalJson(data))
}

/** Recompute the deterministic content hash from a parsed receipt body. */
export function recomputeContentHash(data: ReceiptData): string {
  const body: ReceiptData = {}
  for (const key of Object.keys(data)) {
    if (!IDENTITY_FIELDS.has(key) && key !== 'timestamp_utc') body[key] = data[key]!
  }
  return hashJson(body)
}

/** Recompute the emission identity (includes timestamp_utc). */
export function recomputeReceiptId(data: ReceiptData): string {
  const body: ReceiptData = {}
  for (const key of Object.keys(data)) {
    if (!IDENTITY_FIELDS.has(key)) body[key] = data[key]!
  }
  return hashJson(body)
}

function isPlainObject(value: CanonicalValue): value is ReceiptData {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

/** Recompute the collapsed status from invariant_results, or null if unshaped. */
export function recomputeStatus(data: ReceiptData): ResultStatus | null {
  const invariants = data.invariant_results
  if (!Array.isArray(invariants) || invariants.some((item) => !isPlainObject(item))) {
    return null
  }
  if (data.epistemic_tag === 'FAILED') {
    return 'FAIL'
  }
  const statuses: ResultStatus[] = []
  for (const item of invariants) {
    const raw = (item as ReceiptData).status
    if (typeof raw !== 'string' || !VALID_STATUSES.has(raw)) {
      return null
    }
    statuses.push(raw as ResultStatus)
  }
  return combineStatuses(statuses)
}

function schemaLevel(data: ReceiptData): ValidationLevel {
  const errors: string[] = []
  const missing = [...REQUIRED_RECEIPT_FIELDS].filter((field) => !(field in data)).sort()
  if (missing.length > 0) {
    errors.push(`missing required field(s): ${missing.join(', ')}`)
  }

  if ('receipt_id' in data && !String(data.receipt_id).startsWith('sha256:')) {
    errors.push('receipt_id must start with sha256:')
  }
  if ('content_hash' in data && !String(data.content_hash).startsWith('sha256:')) {
    errors.push('content_hash must start with sha256:')
  }

  if (typeof data.status !== 'string' || !VALID_STATUSES.has(data.status)) {
    errors.push(`status must be one of ${[...VALID_STATUSES].sort().join(', ')}`)
  }

  if (typeof data.epistemic_tag !== 'string' || !VALID_TAGS.has(data.epistemic_tag)) {
    errors.push(`epistemic_tag must be one of ${[...VALID_TAGS].sort().join(', ')}`)
  }

  if ('input_hashes' in data) {
    const inputHashes = data.input_hashes
    if (!Array.isArray(inputHashes)) {
      errors.push('input_hashes must be a list')
    } else if (inputHashes.length === 0) {
      errors.push('input_hashes must be non-empty')
    }
  }

  if ('invariant_results' in data) {
    const invariants = data.invariant_results
    if (!Array.isArray(invariants)) {
      errors.push('invariant_results must be a list')
    } else if (invariants.length === 0) {
      errors.push('invariant_results must contain at least one check; empty invariants never mean PASS')
    } else {
      invariants.forEach((invariant, index) => {
        if (!isPlainObject(invariant)) {
          errors.push(`invariant_results[${index}] must be an object`)
          return
        }
        const status = invariant.status
        if (typeof status !== 'string' || !VALID_STATUSES.has(status)) {
          errors.push(`invariant_results[${index}].status must be PASS, WARN, or FAIL`)
        }
        if (!invariant.name) {
          errors.push(`invariant_results[${index}].name is required`)
        }
      })
    }
  }

  return { name: LEVEL_SCHEMA, ok: errors.length === 0, errors }
}

function hashLevel(data: ReceiptData): ValidationLevel {
  const errors: string[] = []
  let expectedContent: string
  try {
    expectedContent = recomputeContentHash(data)
  } catch (exc) {
    errors.push(`body cannot be canonically hashed: ${(exc as Error).message}`)
    return { name: LEVEL_HASH, ok: false, errors }
  }
  const recordedContent = data.content_hash
  if (recordedContent === undefined || recordedContent === null) {
    errors.push('content_hash is missing; cannot verify body integrity')
  } else if (recordedContent !== expectedContent) {
    errors.push('content_hash mismatch: body was tampered with or not kernel-generated')
  }
  const recordedId = data.receipt_id
  if (recordedId === undefined || recordedId === null) {
    errors.push('receipt_id is missing')
  } else {
    const expectedId = recomputeReceiptId(data)
    if (recordedId !== expectedId) {
      errors.push('receipt_id mismatch: emission identity does not match body + timestamp')
    }
  }
  return { name: LEVEL_HASH, ok: errors.length === 0, errors }
}

function statusLevel(data: ReceiptData): ValidationLevel {
  const recomputed = recomputeStatus(data)
  if (recomputed === null) {
    return {
      name: LEVEL_STATUS,
      ok: false,
      errors: ['invariant_results are not shaped well enough to recompute status'],
    }
  }
  const recorded = data.status
  if (recorded !== recomputed) {
    return {
      name: LEVEL_STATUS,
      ok: false,
      errors: [`status mismatch: recorded ${JSON.stringify(recorded)} but invariant_results collapse to ${JSON.stringify(recomputed)}`],
    }
  }
  return { name: LEVEL_STATUS, ok: true, errors: [] }
}

function versionLevel(data: ReceiptData): ValidationLevel {
  const errors: string[] = []
  const kernelVersion = data.kernel_version
  if (typeof kernelVersion !== 'string' || !ALLOWED_KERNEL_VERSIONS.has(kernelVersion)) {
    errors.push(
      `kernel_version ${JSON.stringify(kernelVersion)} is not in the allowlist ${JSON.stringify([...ALLOWED_KERNEL_VERSIONS].sort())}`,
    )
  }
  const metricVersion = data.metric_version
  if (typeof metricVersion !== 'string' || !ALLOWED_METRIC_VERSIONS.has(metricVersion)) {
    errors.push(
      `metric_version ${JSON.stringify(metricVersion)} is not in the allowlist ${JSON.stringify([...ALLOWED_METRIC_VERSIONS].sort())}`,
    )
  }
  const schemaVersion = data.receipt_schema_version
  if (typeof schemaVersion !== 'string' || !ALLOWED_RECEIPT_SCHEMA_VERSIONS.has(schemaVersion)) {
    errors.push(
      `receipt_schema_version ${JSON.stringify(schemaVersion)} is not in the allowlist ${JSON.stringify([...ALLOWED_RECEIPT_SCHEMA_VERSIONS].sort())}`,
    )
  }
  return { name: LEVEL_VERSION, ok: errors.length === 0, errors }
}

/** Validate a parsed receipt at all four levels (schema/hash/status/version). */
export function validateReceiptData(data: CanonicalValue): ReceiptValidation {
  if (!isPlainObject(data)) {
    return new ReceiptValidation([{ name: LEVEL_SCHEMA, ok: false, errors: ['receipt must be a JSON object'] }])
  }
  return new ReceiptValidation([schemaLevel(data), hashLevel(data), statusLevel(data), versionLevel(data)])
}

File: src/langarian/validation.py
"""Shared receipt validation levels for Langarian Math Workbench v0.3.

Receipt validation is split into four distinct, separately reported levels:

- ``schema``: the receipt is shaped like a Langarian receipt (required fields,
  types, known enum values, at least one invariant result). When present,
  ``timestamp_utc`` must be an ISO-8601 UTC timestamp
  (``YYYY-MM-DDTHH:MM:SS[.ffffff](Z|+HH:MM)``); a non-ISO timestamp fails this
  level on import/validation (the TS ledger quarantines the same condition —
  Python treats it as a schema-level failure so CLI ``validate`` rejects it).
  Passing this level alone is a *shape-only* check and must never be labeled
  "verified".
- ``hash``: ``content_hash`` and ``receipt_id`` are recomputed from the body
  and must match the recorded values (tamper detection).
- ``status``: the recorded ``status`` must equal the status recomputed by
  collapsing ``invariant_results`` (with the empty-invariants => FAIL rule and
  the FAILED epistemic tag override).
- ``version``: ``kernel_version``, ``metric_version``, and
  ``receipt_schema_version`` must be in the current allowlists; older or
  unknown versions are rejected (no silent downgrade acceptance).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .epistemic import EpistemicTag, ResultStatus, combine_statuses
from .receipts import canonical_json
from .version import (
    ALLOWED_KERNEL_VERSIONS,
    ALLOWED_METRIC_VERSIONS,
    ALLOWED_RECEIPT_SCHEMA_VERSIONS,
)
import hashlib
import re

# Same pattern as the TS ledger (web/src/ledger/ledger.ts ISO_8601_UTC):
# ISO-8601 with seconds precision and an explicit UTC offset or Z.
ISO_8601_UTC = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$")

REQUIRED_RECEIPT_FIELDS = {
    "receipt_id",
    "content_hash",
    "kernel_version",
    "metric_version",
    "receipt_schema_version",
    "operator",
    "input_hashes",
    "output_hash",
    "invariant_results",
    "status",
    "epistemic_tag",
}
VALID_STATUSES = {"PASS", "WARN", "FAIL"}
VALID_TAGS = {"FORMAL", "COMPUTED", "MODEL", "INTERPRETIVE", "METAPHOR", "OBSERVED", "FAILED"}

LEVEL_SCHEMA = "schema"
LEVEL_HASH = "hash"
LEVEL_STATUS = "status"
LEVEL_VERSION = "version"
LEVEL_ORDER = (LEVEL_SCHEMA, LEVEL_HASH, LEVEL_STATUS, LEVEL_VERSION)

_IDENTITY_FIELDS = {"receipt_id", "content_hash"}


@dataclass(frozen=True)
class ValidationLevel:
    """Outcome of one validation level."""

    name: str
    ok: bool
    errors: tuple[str, ...] = field(default_factory=tuple)

    def to_dict(self) -> dict[str, Any]:
        return {"name": self.name, "ok": self.ok, "errors": list(self.errors)}


@dataclass(frozen=True)
class ReceiptValidation:
    """Full multi-level validation result for one receipt."""

    levels: tuple[ValidationLevel, ...]

    @property
    def ok(self) -> bool:
        return all(level.ok for level in self.levels)

    @property
    def schema_only_ok(self) -> bool:
        """True when only the schema level passed — NOT verification."""

        by_name = {level.name: level.ok for level in self.levels}
        return by_name.get(LEVEL_SCHEMA, False) and not self.ok

    def level(self, name: str) -> ValidationLevel:
        for level in self.levels:
            if level.name == name:
                return level
        raise KeyError(name)

    def errors_flat(self) -> list[str]:
        return [f"{level.name}: {error}" for level in self.levels for error in level.errors]

    def to_dict(self) -> dict[str, Any]:
        return {"ok": self.ok, "levels": [level.to_dict() for level in self.levels]}


def _hash_json(data: Any) -> str:
    return "sha256:" + hashlib.sha256(canonical_json(data).encode("utf-8")).hexdigest()


def recompute_content_hash(data: dict[str, Any]) -> str:
    """Recompute the deterministic content hash from a parsed receipt body."""

    body = {key: value for key, value in data.items() if key not in _IDENTITY_FIELDS and key != "timestamp_utc"}
    return _hash_json(body)


def recompute_receipt_id(data: dict[str, Any]) -> str:
    """Recompute the emission identity (includes timestamp_utc)."""

    body = {key: value for key, value in data.items() if key not in _IDENTITY_FIELDS}
    return _hash_json(body)


def recompute_status(data: dict[str, Any]) -> str | None:
    """Recompute the collapsed status from invariant_results, or None if unshaped."""

    invariants = data.get("invariant_results")
    if not isinstance(invariants, list) or any(not isinstance(item, dict) for item in invariants):
        return None
    if data.get("epistemic_tag") == EpistemicTag.FAILED.value:
        return ResultStatus.FAIL.value
    statuses: list[ResultStatus] = []
    for item in invariants:
        raw = item.get("status")
        if raw not in VALID_STATUSES:
            return None
        statuses.append(ResultStatus(raw))
    return combine_statuses(statuses).value


def _schema_level(data: dict[str, Any]) -> ValidationLevel:
    errors: list[str] = []
    missing = sorted(REQUIRED_RECEIPT_FIELDS - set(data))
    if missing:
        errors.append(f"missing required field(s): {', '.join(missing)}")

    if "receipt_id" in data and not str(data["receipt_id"]).startswith("sha256:"):
        errors.append("receipt_id must start with sha256:")
    if "content_hash" in data and not str(data["content_hash"]).startswith("sha256:"):
        errors.append("content_hash must start with sha256:")

    if data.get("status") not in VALID_STATUSES:
        errors.append(f"status must be one of {sorted(VALID_STATUSES)}")

    if data.get("epistemic_tag") not in VALID_TAGS:
        errors.append(f"epistemic_tag must be one of {sorted(VALID_TAGS)}")

    if "timestamp_utc" in data:
        timestamp = data["timestamp_utc"]
        if not isinstance(timestamp, str) or not ISO_8601_UTC.match(timestamp):
            errors.append(
                "timestamp_utc must be ISO-8601 UTC format "
                "(YYYY-MM-DDTHH:MM:SS[.ffffff](Z|+HH:MM)); "
                "ledger ordering by time is unverified"
            )

    input_hashes = data.get("input_hashes")
    if "input_hashes" in data:
        if not isinstance(input_hashes, list):
            errors.append("input_hashes must be a list")
        elif not input_hashes:
            errors.append("input_hashes must be non-empty")

    invariants = data.get("invariant_results")
    if "invariant_results" in data:
        if not isinstance(invariants, list):
            errors.append("invariant_results must be a list")
        elif not invariants:
            errors.append("invariant_results must contain at least one check; empty invariants never mean PASS")
        else:
            for index, invariant in enumerate(invariants):
                if not isinstance(invariant, dict):
                    errors.append(f"invariant_results[{index}] must be an object")
                    continue
                if invariant.get("status") not in VALID_STATUSES:
                    errors.append(f"invariant_results[{index}].status must be PASS, WARN, or FAIL")
                if not invariant.get("name"):
                    errors.append(f"invariant_results[{index}].name is required")

    return ValidationLevel(LEVEL_SCHEMA, not errors, tuple(errors))


def _hash_level(data: dict[str, Any]) -> ValidationLevel:
    errors: list[str] = []
    try:
        expected_content = recompute_content_hash(data)
    except (TypeError, ValueError) as exc:
        errors.append(f"body cannot be canonically hashed: {exc}")
        return ValidationLevel(LEVEL_HASH, False, tuple(errors))
    recorded_content = data.get("content_hash")
    if recorded_content is None:
        errors.append("content_hash is missing; cannot verify body integrity")
    elif recorded_content != expected_content:
        errors.append("content_hash mismatch: body was tampered with or not kernel-generated")
    recorded_id = data.get("receipt_id")
    if recorded_id is None:
        errors.append("receipt_id is missing")
    else:
        expected_id = recompute_receipt_id(data)
        if recorded_id != expected_id:
            errors.append("receipt_id mismatch: emission identity does not match body + timestamp")
    return ValidationLevel(LEVEL_HASH, not errors, tuple(errors))


def _status_level(data: dict[str, Any]) -> ValidationLevel:
    recomputed = recompute_status(data)
    if recomputed is None:
        return ValidationLevel(LEVEL_STATUS, False, ("invariant_results are not shaped well enough to recompute status",))
    recorded = data.get("status")
    if recorded != recomputed:
        return ValidationLevel(
            LEVEL_STATUS,
            False,
            (f"status mismatch: recorded {recorded!r} but invariant_results collapse to {recomputed!r}",),
        )
    return ValidationLevel(LEVEL_STATUS, True, ())


def _version_level(data: dict[str, Any]) -> ValidationLevel:
    errors: list[str] = []
    kernel_version = data.get("kernel_version")
    if kernel_version not in ALLOWED_KERNEL_VERSIONS:
        errors.append(f"kernel_version {kernel_version!r} is not in the allowlist {sorted(ALLOWED_KERNEL_VERSIONS)}")
    metric_version = data.get("metric_version")
    if metric_version not in ALLOWED_METRIC_VERSIONS:
        errors.append(f"metric_version {metric_version!r} is not in the allowlist {sorted(ALLOWED_METRIC_VERSIONS)}")
    schema_version = data.get("receipt_schema_version")
    if schema_version not in ALLOWED_RECEIPT_SCHEMA_VERSIONS:
        errors.append(
            f"receipt_schema_version {schema_version!r} is not in the allowlist {sorted(ALLOWED_RECEIPT_SCHEMA_VERSIONS)}"
        )
    return ValidationLevel(LEVEL_VERSION, not errors, tuple(errors))


def validate_receipt_data(data: dict[str, Any]) -> ReceiptValidation:
    """Validate a parsed receipt at all four levels (schema/hash/status/version).

    Note: even full success is local consistency verification, not
    recomputation of the underlying mathematical operation.
    """

    if not isinstance(data, dict):
        level = ValidationLevel(LEVEL_SCHEMA, False, ("receipt must be a JSON object",))
        return ReceiptValidation((level,))
    return ReceiptValidation(
        (
            _schema_level(data),
            _hash_level(data),
            _status_level(data),
            _version_level(data),
        )
    )


__all__ = [
    "ISO_8601_UTC",
    "REQUIRED_RECEIPT_FIELDS",
    "VALID_STATUSES",
    "VALID_TAGS",
    "LEVEL_SCHEMA",
    "LEVEL_HASH",
    "LEVEL_STATUS",
    "LEVEL_VERSION",
    "LEVEL_ORDER",
    "ValidationLevel",
    "ReceiptValidation",
    "recompute_content_hash",
    "recompute_receipt_id",
    "recompute_status",
    "validate_receipt_data",
]
